src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const stateCopy = {};
+
+  Object.assign(stateCopy, state);
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      for (const key in action.extraData) {
+        stateCopy[key] = action.extraData[key];
+      }
+
+      stateHistory.push(stateCopy);
+    } else if (action.type === 'removeProperties') {
+      for (const key in action.keysToRemove) {
+        delete stateCopy[key];
+      }
+
+      stateHistory.push(stateCopy);
```

Comment:
Similar to line 21, ensure that a new copy of `stateCopy` is created before pushing it to `stateHistory` to avoid all entries being the same object.